# SCIPhI: Single-cell mutation identification via phylogenetic inference
#
# Copyright (C) 2018 ETH Zurich, Jochen Singer
#
# This file is part of SCIPhI.
#
# SCIPhIN is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# SCIPhIN is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with SCIPhI. If not, see <http://www.gnu.org/licenses/>.
# 
# @author: Jochen Singer
import os
from subprocess import call
from snakemake.utils import R
import random

OUTDIR = ""
NUMPOSITIONS = 1000000
DROPOUTRATE = 0.2
LOOPS = 1000000
MAXCOVERAGE = 300
NUMMUTATIONS = 100
NUMCELLS = ['10', '25', '50']
PRIORSSCIPHI = ['0.01','0.001','0.0001','0.00001','0.000001',]
PRIORSMONOVAR = ['0.2','0.02','0.002','0.0002','0.00002',]
DROPOUTS = ['0', '0.1', '0.2', '0.3']
ZYGS = ['0', '0.1', '0.2', '0.3']
CPNS = ['0', '0.125', '0.25', '0.375', '0.5']
NUMRUNS = 50
MDAERRORRATE = 0.0000005
SEQUENCINGERRORRATE = 0.001
COVMEAN = 25
COVVAR = 50
AVERAGEREGIONLENGTH = 1000
ALPHA = 100
BETA = 100
VIOLATIONS = ['0','0.05','0.10']

include: "tools.py"
include: "createSimulationDataSets.snake"

rule all:
    input:
        OUTDIR + 'distance_numCells.txt',
        OUTDIR + 'noNAs_distance_numCells.txt',
        
